core/trino-parser/src/main/java/io/trino/sql/jsonpath/PathParser.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.jsonpath;
+
+import io.trino.jsonpath.JsonPathBaseListener;
+import io.trino.jsonpath.JsonPathLexer;
+import io.trino.jsonpath.JsonPathParser;
+import io.trino.sql.jsonpath.tree.PathNode;
+import io.trino.sql.parser.ParsingException;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonToken;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.atn.PredictionMode;
+import org.antlr.v4.runtime.misc.Pair;
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public final class PathParser
+{
+    private final BaseErrorListener errorListener;
+
+    public PathParser(Location startLocation)
+    {
+        requireNonNull(startLocation, "startLocation is null");
+
+        int pathStartLine = startLocation.line;
+        int pathStartColumn = startLocation.column;
+        this.errorListener = new BaseErrorListener()
+        {
+            @Override
+            public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String message, RecognitionException e)
+            {
+                // re-position the exception relatively to the start of the path expression within the query
+                int lineInQuery = pathStartLine - 1 + line;
+                int columnInQuery = line == 1 ? pathStartColumn + 1 + charPositionInLine : charPositionInLine + 1;
```

Comment:
if the query is formatted across multiple lines like this (`XXXX` is the path):
```
=================================
================XXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXX=========================
===========
```
the path parser sees only this, as if the path started at first line, first column:
```
XXXXXXXXXXXXXX
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
XXXXXX
```
If there is parsing exception, the location is aligned relatively to what the parser sees.

After the parsing exception, the whole query is printed, so we must adjust the error location so that the error highlight matches. It involves skipping the lines of the query preceding the line where the path starts, and also skipping the initial characters of the first line containing the path, if the error was in this line.

```
SELECT json_exists(
'[0, 1, 2, 3
, 4, 5, 6]',
  'lax
$[10]
          [blabla]'
)
```
```
Query 20220405_102431_00004_qk4zw failed: line 6:12
```

Do you think I should add more comment in the code?

Reply:
> Do you think I should add more comment in the code?

Thanks maybe a bit. Sth like this?
```
                // The line and charPositionInLine correspond to the character within the string literal with JSON path expression.
                // Line and offset in error returned to the user should be computed based on the beginning of the whole query text.
                // We re-position the exception relatively to the start of the path expression within the query.
```
More verbose; but could help some people (like me)